pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java

```diff
@@ -231,185 +197,149 @@ public void initialize(ServiceConfiguration conf, PulsarResources pulsarResource
-                        log.info("[{}] Successfully granted access for role {}: {} - topic {}", role, role, actions,
-                                topicUri);
-                    }
-                });
+        return getPoliciesReadOnlyAsync().thenCompose(readonly -> {
+            if (readonly) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Policies are read-only. Broker cannot do read-write operations");
+                }
+                throw new IllegalStateException("policies are in readonly mode");
+            }
+            String topicUri = topicName.toString();
+            return pulsarResources.getNamespaceResources()
+                    .setPoliciesAsync(topicName.getNamespaceObject(), policies -> {
+                        policies.auth_policies.getTopicAuthentication()
+                                .computeIfAbsent(topicUri, __ -> new HashMap<>())
+                                .put(role, actions);
+                        return policies;
+                    }).whenComplete((__, ex) -> {
+                        if (ex != null) {
+                            log.error("Failed to set permissions for role {} on topic {}", role, topicName, ex);
+                        } else {
+                            log.info("Successfully granted access for role {}: {} - topic {}", role, actions, topicUri);
+                        }
+                    });
+        });
     }
 
     @Override
     public CompletableFuture<Void> grantPermissionAsync(NamespaceName namespaceName, Set<AuthAction> actions,
-            String role, String authDataJson) {
-        try {
-            validatePoliciesReadOnlyAccess();
-        } catch (Exception e) {
-            return FutureUtil.failedFuture(e);
-        }
-
-        return pulsarResources.getNamespaceResources()
-                .setPoliciesAsync(namespaceName, policies -> {
-                    policies.auth_policies.getNamespaceAuthentication().put(role, actions);
-                    return policies;
-                }).whenComplete((__, throwable) -> {
-                    if (throwable != null) {
-                        log.error("[{}] Failed to set permissions for role {} namespace {}", role, role, namespaceName,
-                                throwable);
-                    } else {
-                        log.info("[{}] Successfully granted access for role {}: {} - namespace {}", role, role, actions,
-                                namespaceName);
-                    }
-                });
+                                                        String role, String authDataJson) {
+        return getPoliciesReadOnlyAsync().thenCompose(readonly -> {
+            if (readonly) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Policies are read-only. Broker cannot do read-write operations");
+                }
+                throw new IllegalStateException("policies are in readonly mode");
```

Comment:
If no one catches the exception, the future will never be complete. For example:

```
        CompletableFuture future = new CompletableFuture();
        future.thenRun(() -> {
            System.out.println("first thing");
        }).thenCompose(b -> {
            throw new IllegalArgumentException("fail future");
        }).thenRun(() -> {
            System.out.println("other things");
        });
        future.get();
```